packages/forms/src/model.ts

```diff
@@ -2119,11 +2126,12 @@ export class FormArray extends AbstractControl {
    * replaced with a new one. When false, no events are emitted.
    */
   setControl(index: number, control: AbstractControl, options: {emitEvent?: boolean} = {}): void {
-    if (this.controls[index]) this.controls[index]._registerOnCollectionChange(() => {});
-    this.controls.splice(index, 1);
+    const adjustedIdx = this._adjustIndex(index);
```

Comment:
I was looking at this change more and while having `const adjustedIdx = this._adjustIndex(index);` makes sense in a simple case (like `at`), there is a chance that both indices can be used in the code by accident. In order to avoid this, we can reuse the `index` var and set it to the adjusted index as a first line in affected functions:
```suggestion
    index = this._adjustIndex(index);
```
What do you think?

Reply:
I actually prefer your suggestion, and this is what I've done originally before implementing and using `_adjustIndex()`. I took inspiration from this code:
https://github.com/angular/angular/blob/d0246d1d0a4aee420c0c85e8f31a6b24fb982b1f/packages/core/src/linker/view_container_ref.ts#L393